### PR TITLE
feat: Address to Remove page

### PR DESCRIPTION
**Pull request by rhysbarrett**
### JIRA link

https://companieshouse.atlassian.net/browse/SR-26

### Change description

This commit implements and tests the "Address to Remove" page. It does not include navigation.

### Page Screenshot (if applicable)

**Page:**

**Page with Errors:**

### Work Checklist

- [x] Tests added where applicable
- [ ] UI changes look good on mobile
- [ ] UI changes meet accessibility criteria

---

### Merge Instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/models/Address.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export interface Address {
+  addressLine1: string,
```
Comment (n-r-b) [inline src/models/Address.ts:2]: can we simplify all of these and remove address from the property names? for example, `line1`, `line2`

File: src/validators/schema/AddressToRemoveSchema.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import Joi from 'joi';
+
+import { basicString } from './BasicStringSchemaItem'
+
+export const schema = Joi.object({
+  addressLine1: basicString('Building and street is required'),
+  addressLine2: Joi.string().allow(null, ''),
```
Comment (n-r-b) [inline src/validators/schema/AddressToRemoveSchema.ts:7]: let's say this is not null or empty, should we still validate the string with a pattern?

File: src/views/address-to-remove.njk [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+{% from 'govuk/components/error-summary/macro.njk' import govukErrorSummary %}
+{% from 'govuk/components/input/macro.njk'         import govukInput %}
+{% from 'govuk/components/button/macro.njk'        import govukButton %}
+{% from "govuk/components/fieldset/macro.njk"      import govukFieldset %}
+
+{% block pageTitle %}
+  Address details
+{% endblock %}
+
+{% set legendText %}
```
Comment (n-r-b) [inline src/views/address-to-remove.njk:12]: Where is this used?

File: test/validators/schema/AddressToRemoveSchema.test.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        addressPostcode: 'CF14 3UZ'
+      }
+    }
+
+    it('should allow non-empty values for all fields', () => {
+      const validationResult = validator.validate(generateTestData());
+      assertValidationErrors(validationResult, []);
+    });
+
+    it('should allow empty value for Address Line 2', () => {
```
Comment (n-r-b) [inline test/validators/schema/AddressToRemoveSchema.test.ts:77]: do we have a test for allowing a null line2?

File: src/views/address-to-remove.njk (hunk 3 above)
Comment (rhysbarrett) [reply src/views/address-to-remove.njk:12]: Oh, it isn't, good spot. GOV.UK's built-in fieldset is missing the ability to include a top-level hint, so I was preparing to do some hacking, but in the end I opted for just using plain ol' HTML.

File: src/controllers/AddressToRemoveController.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  public processForm = async (req: Request, res: Response, next: NextFunction) => {
+    const validationResult: ValidationResult = await this.validator.validate(req);
+    if (validationResult.errors.length > 0) {
+      res.status(StatusCodes.UNPROCESSABLE_ENTITY);
+      return res.render(template, {
+        ...req.body,
+        validationResult
+      });
+    } else {
+      const session = SessionService.getSuppressionSession(req);
```
Comment (n-r-b) [inline src/controllers/AddressToRemoveController.ts:31]: have you created tests to cover this session logic in the controller?
- rendering the page when there's data in the session
- throwing exception on POST when session data doesn't exist

File: test/validators/schema/AddressToRemoveSchema.test.ts (hunk 4 above)
Comment (rhysbarrett) [reply test/validators/schema/AddressToRemoveSchema.test.ts:77]: We do now :)

File: src/controllers/AddressToRemoveController.ts (hunk 5 above)
Comment (rhysbarrett) [reply src/controllers/AddressToRemoveController.ts]: I've just finished wrestling with the test for the POST error (turns out we don't yet have a default error handler, so just throwing an Exception in the controller doesn't cut it). I'll take a look at the session rendering now.

File: src/validators/schema/AddressToRemoveSchema.ts (hunk 2 above)
Comment (rhysbarrett) [reply src/validators/schema/AddressToRemoveSchema.ts]: Potentially. I'm not sure it's worth the effort at this stage though, as it doesn't seem to add any value. " " is functionally equivalent to a missing or empty value in this context. Also, it's worth considering the frustration that would ensue if:
1. A user accidentally adds a space to Line 2 while tabbing through the fields
2. We then render a validation error for that field
3. User cannot see anything wrong with the field, as the field only contains whitespace
4. User then clears the field, just in case
5. All of a sudden the details are accepted, with no explanation about what went wrong before.

We can't use the standard validation message "Address Line 2 is required", because that's not true.

File: src/controllers/AddressToRemoveController.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        return res.render(template, {
+          ...req.body,
+          validationResult
+        });
+      } else {
+        session.addressToRemove = req.body as Address;
+        SessionService.setSuppressionSession(req, session);
+        res.redirect(ADDRESS_TO_REMOVE_PAGE_URI);
+      }
+    }
```
Comment (akaJSaunders) [inline src/controllers/AddressToRemoveController.ts:39]: we could prob decrease some of this nesting, what do you think

```suggestion
    if (!session) {
      res.status(StatusCodes.INTERNAL_SERVER_ERROR).render('error');
    } 
      
    const validationResult: ValidationResult = await this.validator.validate(req);
    if (validationResult.errors.length > 0) {
      res.status(StatusCodes.UNPROCESSABLE_ENTITY);
      return res.render(template, {
        ...req.body,
        validationResult
      });
    } else {
      session.addressToRemove = req.body as Address;
      SessionService.setSuppressionSession(req, session);
      res.redirect(ADDRESS_TO_REMOVE_PAGE_URI);
      }
    }
```

File: src/routes/routes.ts [hunk 7]
```diff
@@ -19,3 +25,5 @@ routes.get(ROOT_URI, startPageController.renderView);
 routes.post(ROOT_URI, startPageController.start);
 routes.get(APPLICANT_DETAILS_PAGE_URI, applicantDetailsController.renderView);
 routes.post(APPLICANT_DETAILS_PAGE_URI, applicantDetailsController.processForm);
+routes.get(ADDRESS_TO_REMOVE_PAGE_URI, addressToRemoveController.renderView);
+routes.post(ADDRESS_TO_REMOVE_PAGE_URI, addressToRemoveController.processForm);
```
Comment (akaJSaunders) [inline src/routes/routes.ts:29]: I'd prefer if we had spaces between each page for readability. If we were to use Inversify we could probably avoid this whole routes list, but we can discuss that in the upcoming meetings

```suggestion
routes.post(ROOT_URI, startPageController.start);

routes.get(APPLICANT_DETAILS_PAGE_URI, applicantDetailsController.renderView);
routes.post(APPLICANT_DETAILS_PAGE_URI, applicantDetailsController.processForm);

routes.get(ADDRESS_TO_REMOVE_PAGE_URI, addressToRemoveController.renderView);
routes.post(ADDRESS_TO_REMOVE_PAGE_URI, addressToRemoveController.processForm);
```

File: src/validators/schema/AddressToRemoveSchema.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+export const schema = Joi.object({
+  line1: basicString('Building and street is required'),
+  line2: Joi.string().allow(null, ''),
+  town: basicString('Town or city is required'),
+  county: basicString('County is required'),
+  postcode: basicString('Postcode is required')
+});
+
+
```
Comment (akaJSaunders) [inline src/validators/schema/AddressToRemoveSchema.ts:13]: dont mind me
```suggestion
```

File: src/validators/schema/BasicStringSchemaItem.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import Joi from 'joi';
+
+export function basicString(errorMessage: string): Joi.StringSchema {
+  return Joi.string()
+    .required()
+    .pattern(/\w+/)
+    .messages({
+      'any.required': errorMessage,
+      'string.base': errorMessage,
+      'string.empty': errorMessage,
```
Comment (akaJSaunders) [inline src/validators/schema/BasicStringSchemaItem.ts:10]: what is the pattern testing for and does one of these .empty, .base or .required cover it already?

File: src/controllers/AddressToRemoveController.ts (hunk 6 above)
Comment (rhysbarrett) [reply src/controllers/AddressToRemoveController.ts:39]: I did initially try this but res.status() isn't a terminal. I'll have a fiddle with returns and see if I can get it to work.

File: src/views/error.njk [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        <b>+44 0303 1234 500</b>
+      </p>
+      <p>
+        Opening times: <br/>
+        <b>Monday to Friday: 8.30am to 6pm</b>
+      </p>
+      <p class="govuk-body">Closed Easter Sunday, Christmas Day, Boxing Day and New Year`s Day.</p>
+    </div>
+  </div>
+{% endblock %}
```
Comment (akaJSaunders) [inline src/views/error.njk:33]: Out of scope

File: src/controllers/AddressToRemoveController.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+  public renderView = (req: Request, res: Response, next: NextFunction) => {
+    const session = SessionService.getSuppressionSession(req);
+    res.render(template, { ...session?.addressToRemove });
+  }
+
+  public processForm = async (req: Request, res: Response, next: NextFunction) => {
+    const session = SessionService.getSuppressionSession(req);
+    if (!session) {
+      res.status(StatusCodes.INTERNAL_SERVER_ERROR).render('error');
```
Comment (akaJSaunders) [inline src/controllers/AddressToRemoveController.ts:25]: this is out of scope, ticket does not specify an error page

File: src/controllers/AddressToRemoveController.ts (hunk 6 above)
Comment (akaJSaunders) [reply src/controllers/AddressToRemoveController.ts:39]: I comment this later, but this error template is out of scope anyway

File: src/validators/schema/BasicStringSchemaItem.ts (hunk 9 above)
Comment (rhysbarrett) [reply src/validators/schema/BasicStringSchemaItem.ts:10]: It tests for at least one non-whitespace character, preventing the user from submitting, say, a space. 

`.empty` only covers the empty string "", `.base` enforces the string type and `.required` does what it says on the tin.

File: src/views/error.njk (hunk 10 above)
Comment (rhysbarrett) [reply src/views/error.njk:33]: Functionality for the Address to Remove page requires session to be present. If session isn't present, an error needs to be thrown. As we don't yet have a default error handler, throwing an error would lead to an unhandled exception, crashing the application and failing the unit test. To resolve this, an error page needs to be rendered.

The error page might not be in scope but it's a dependency of this story.

File: src/controllers/AddressToRemoveController.ts (hunk 11 above)
Comment (rhysbarrett) [reply src/controllers/AddressToRemoveController.ts:25]: See previous comment. Ticket does not specify error page, but this is a technical requirement. We cannot throw unhandled exceptions in a web application.

File: src/controllers/AddressToRemoveController.ts (hunk 6 above)
Comment (rhysbarrett) [reply src/controllers/AddressToRemoveController.ts:39]: Responded to the comments about the error page. Managed to decrease nesting by returning the `res.render...`.

File: src/views/error.njk (hunk 10 above)
Comment (akaJSaunders) [reply src/views/error.njk:33]: Ok, I would then remove the content from the error page and we simply don't know what Companies House might want on the page.

Comment (rhysbarrett) [reply src/views/error.njk:33]: Okie dokie, I'll strip the content and leave only the title and page heading.

File: src/controllers/AddressToRemoveController.ts (hunk 5 above)
Comment (rhysbarrett) [reply src/controllers/AddressToRemoveController.ts]: Done!

File: src/validators/schema/AddressToRemoveSchema.ts (hunk 2 above)
Comment (n-r-b) [reply src/validators/schema/AddressToRemoveSchema.ts]: ok, that is a fair point 👍

**Review by akaJSaunders — APPROVED** [review]
